### PR TITLE
DAOS-7732 dfuse: Provide a fuse logging callback.

### PR DESCRIPTION
**Pull request by ashleypittman**
When available provide a fuse logging callback to capture
fuse logging in the daos log.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/client/dfuse/SConscript [hunk 1]
```diff
@@ -151,6 +151,9 @@ def configure_fuse(cenv):
                                'cache_readdir'):
         cenv.AppendUnique(CPPDEFINES=['-DHAVE_CACHE_READDIR=1'])
 
+    if check.CheckHeader('fuse3/fuse_log.h'):
+        cenv.AppendUnique(CPPDEFINES=['-DHAVE_FUSE_LOG=1'])
```
Comment (jolivier23) [inline src/client/dfuse/SConscript:155]: Didn't realize you could do this without defining the 0 case but my local test doesn't issue a warning.   Maybe there was some compiler in my past that complained about an undefined macro....or maybe I'm just dreaming.

File: src/client/dfuse/dfuse_core.c [hunk 2]
```diff
@@ -885,6 +889,47 @@ dfuse_ie_close(struct dfuse_projection_info *fs_handle,
...
+		d_level = DLOG_WARN;
+		break;
+	case FUSE_LOG_NOTICE:
+		d_level = DLOG_NOTE;
+		break;
+	case FUSE_LOG_INFO:
+		d_level = DLOG_INFO;
+		break;
+	case FUSE_LOG_DEBUG:
+		d_level = DLOG_DBG;
```
Comment (jolivier23) [inline src/client/dfuse/dfuse_core.c:924]: I think you just want DB_IO here.   DLOG_DBG is a bit mask that covers all debug bits so this is equivalent to something like

0xff | 0x1 where DLOG_DBG is 0xff and DB_IO is 0x1 (not actual values).

File: src/client/dfuse/dfuse_core.c [hunk 3]
```diff
@@ -885,6 +889,47 @@ dfuse_ie_close(struct dfuse_projection_info *fs_handle,
...
+		break;
+	}
+
+	mask = d_log_check(d_level | DB_IO | DD_FAC(fuse));
```
Comment (jolivier23) [inline src/client/dfuse/dfuse_core.c:928]: maybe it doesn't matter much but you lose some efficiency of macros here.  The D_ macro creates a single variable that it checks (and caches the result) so you essentially bypass subsequent calls to d_log_check.

I think maybe what you want here is something like this:

int *flag = &DD_FLAG(DLOG_EMIT, D_LOGFAC);

with the switch also grabbing the right variable.

Then it's a matter of mimicking _D_DEBUG macro...so something like

if (__builtin_expect(*flag, 0)) {
     if (*flag == DLOG_UNINIT) {
          _D_LOG_CHECK(*flag, *flag | D_LOGFAC, ##__VA_ARGS__);
          break;
     }
_D_LOG_NOCHECK(*flag, ##__VA_ARGS__);

So maybe we need another macro here that takes a variable as an argument...and modify the existing calls to _D_DEBUG to call it with DD_FLAG result as an argument

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/client/dfuse/dfuse_core.c [hunk 4]
```diff
@@ -1049,6 +1053,45 @@ dfuse_ie_close(struct dfuse_projection_info *fs_handle,
...
+		d_level = DLOG_WARN;
+		break;
+	case FUSE_LOG_NOTICE:
+		d_level = DLOG_NOTE;
+		break;
+	case FUSE_LOG_INFO:
+		d_level = DLOG_INFO;
+		break;
+	case FUSE_LOG_DEBUG:
+		d_level = DLOG_DBG;
```
Comment (jolivier23) [inline src/client/dfuse/dfuse_core.c:1086]: Put DB_IO here and remove the DB_IO below.

DLOG_DBG is 0x00ffff00 so it sets all debug streams (controlled by DD_MASK)

DLOG_IO is only one of those bits which allows one to turn it off even if capturing other debug.

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Comment by ashleypittman** [comment]
It is a draft.

**Pull request closed without merging**